crates/gpui/src/arena.rs

```diff
@@ -79,30 +121,53 @@ impl Arena {
 
         unsafe {
             let layout = alloc::Layout::new::<T>();
-            let offset = self.offset.add(self.offset.align_offset(layout.align()));
-            let next_offset = offset.add(layout.size());
-            assert!(next_offset <= self.end, "not enough space in Arena");
-
-            let result = ArenaBox {
-                ptr: offset.cast(),
-                valid: self.valid.clone(),
+            let mut current_chunk = self.chunks.get_mut(self.current_chunk).unwrap();
+            let ptr = if let Some(ptr) = current_chunk.allocate(layout) {
+                ptr
+            } else if self.current_chunk + 1 < self.chunks.len()
```

Comment:
2 points, but please ignore if they seem like a nitpick, I'm really not sure how important they are, if at all:
* Would be nice to add a test case for this scenario.
* In some situations we may end up with an unexpected (?) behaviour if we need to skip a chunk. Example:
  Arena after growing and getting cleared: [1 MB] -> [1 MB] -> [1 MB] -> [1MB]
  Now we need to allocate 2MB: [1 MB] -> [1 MB] -> [1 MB] -> [1MB] -> [2 MB]
  All the 1MB chunks are left unused.